File: readmefetch/src/draw_ascii.ts
import { createCanvas, loadImage } from "@napi-rs/canvas";

const ASCII_CHARS = '.:-=+*#%@"';

function getAsciiChar(pixel: [number, number, number]): string {
  const brightness = (pixel[0] + pixel[1] + pixel[2]) / 3;
  const charIndex = Math.min(
    Math.floor((brightness / 255) * (ASCII_CHARS.length - 1)),
    ASCII_CHARS.length - 1,
  );
  return ASCII_CHARS[charIndex];
}

async function imageToAscii(url: string, width = 50): Promise<string> {
  const response = await fetch(url);
  const buffer = Buffer.from(await response.arrayBuffer());
  const img = await loadImage(buffer);
  const aspectRatio = img.width / img.height;
  const height = Math.floor(width * aspectRatio * 0.5);

  const canvas = createCanvas(width, height);
  const ctx = canvas.getContext("2d");
  ctx.drawImage(img, 0, 0, width, height);

  const imageData = ctx.getImageData(0, 0, width, height);
  const pixels = imageData.data;

  let asciiStr = "";
  for (let y = 0; y < height; y++) {
    for (let x = 0; x < width; x++) {
      const idx = (y * width + x) * 4;
      const pixel: [number, number, number] = [pixels[idx], pixels[idx + 1], pixels[idx + 2]];
      asciiStr += getAsciiChar(pixel);
    }
    asciiStr += "\n";
  }

  return asciiStr;
}

export { imageToAscii };

File: readmefetch/src/draw_ascii.py
import requests
from github import Github
from PIL import Image
from io import BytesIO
from src.fetch_info import fetch_stats

def get_ascii_char(pixel):
    ascii_chars = '.:-=+*#%@"'
    brightness = sum(pixel)/3
    char_index = min(int(brightness/255 * (len(ascii_chars)-1)), len(ascii_chars)-1)
    return ascii_chars[char_index]

def image_to_ascii(image, width=50) -> str:
    aspect_ratio = image.width / image.height
    height = int((width*aspect_ratio)*0.5)

    image = image.resize((width, height))
    image = image.convert('RGB')
    ascii_str = ""

    for y in range(height):
        for x in range(width):
            pixel = image.getpixel((x,y))
            ascii_str += get_ascii_char(pixel)
        ascii_str += "\n"

    return ascii_str

def generate_logo(g:Github) -> str:
    user_pfp = g.get_user().avatar_url
    response = requests.get(user_pfp)
    img = Image.open(BytesIO(response.content))

    return image_to_ascii(img)
